---
title: "Microbiome and obesity in Children of 1997 birth cohort"
author: "Baoting HE"
date: "2024-11-27"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{microbiome-and-obesity-in-BC}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include=F}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
options(rmarkdown.html_vignette.check_title = FALSE)
```

```{r setup, include=FALSE}
install.packages("D:/Annabel/OneDrive - The University Of Hong Kong/Study/For RAP/R/R code_github/micro97BCpkg_0.0.0.9000.tar.gz", repos = NULL, type = "source")
library(micro97BCpkg)
library(devtools)
library(data.table)
library(pairwiseAdonis)
library("ggbiplot")
library(beeswarm)
library(ggpubr)
library(ggbeeswarm)
library(readr)
library("dplyr")
library(ape)
library(vegan)
library(LDM)  ##### analyzing matched-set data using the LDM or PERMANOVA 
library(stringr)
library(survival)

```

# Introduction

This is the source code for the study in microbiome and obesity.

# Step 1: Data preparation

The study is based on the data obtained from "Children of 1997" Hong Kong Chinese birth cohort. Based on the ethic approval of this study, research data is not publicly shared currently.

**Data preparation and data cleaning for microbiome abundance data**

```{r data preparation, warning=FALSE, include=FALSE}
##Read the data
#Read failed samples, based on multiqc report
failed_data <- read.csv("D:/Annabel/OneDrive - The University Of Hong Kong/Study/postdoc/coprojects/Microbiome and obesity/failed_samples.csv")
failed_sample <- failed_data$Per.Sequence.GC.Content
failed_sample <-  gsub("_good.*$","",failed_sample)
failed_sample <- failed_sample[!duplicated(failed_sample)]


#Read metaphlan data

metaphlan_abundance <- fread("D:/Annabel/OneDrive - The University Of Hong Kong/Study/postdoc/coprojects/Microbiome and obesity/merged_abundance_table_mpa4.txt")
metaphlan_abundance <- as.data.frame(metaphlan_abundance)

rownames(metaphlan_abundance) <- metaphlan_abundance$clade_name; metaphlan_abundance <- metaphlan_abundance[,-1]

#data cleaning
metaphlan_species <- metaphlan_abundance[grep("s__",row.names(metaphlan_abundance)),]
metaphlan_species <- metaphlan_species[-grep("t__",row.names(metaphlan_species)),]
row.names(metaphlan_species) <- gsub("^.*s__","",row.names(metaphlan_species))
metaphlan_species <- metaphlan_species[,grep("BC",colnames(metaphlan_species))]
metaphlan_species_all <- metaphlan_species   ## all species (no filter applied)

#Read clinical data

clinical_data <- read.csv("D:/Annabel/OneDrive - The University Of Hong Kong/Study/postdoc/coprojects/Microbiome and obesity/97BC_Microbiome Data_20220915.csv",header = T)
clinical_data <- clinical_data[,-1]
colnames(clinical_data)


##Data cleaning procedure
#exlude the pairs with failed samples and unpair sex
clinical_data <- clinical_data[-which(clinical_data$Lab.ID %in% failed_sample),]
match<-as.numeric(names(which(table(clinical_data$subclass)==2)))
unmatch<-as.numeric(names(which(table(clinical_data$subclass)==1)))
unmatch.id<-clinical_data[(clinical_data$subclass %in% unmatch),]$Lab.ID
clinical_data <- clinical_data[order(clinical_data$subclass),]
for (sample in 1:(nrow(clinical_data)-1)) {
  if (clinical_data[sample,"sex"]!=clinical_data[sample+1,"sex"] && clinical_data[sample,"subclass"]==clinical_data[sample+1,"subclass"]) {
    unmatch.id <- c(unmatch.id,clinical_data$Lab.ID[sample],clinical_data$Lab.ID[sample+1])
  }
}
metaphlan_species<- metaphlan_species[,-which(colnames(metaphlan_species) %in% c(failed_sample,unmatch.id))]
#exclude the species with 0 abundance in all samples
metaphlan_species <- metaphlan_species[-c(which(rowSums(metaphlan_species)==0)),]

```

**Plot samples by counts of species over 0.05 abundance**

```{r plot to see the distribution of different counts of species, fig.height=5, fig.width=7, warning=FALSE}
sample_prevalence <- as.data.frame(rowSums(metaphlan_species>0.05))

colnames(sample_prevalence) <- "Number_of_samples"


plot_sample_prevalence<-sample_prevalence %>% 
  mutate(Number_of_samples = ifelse(Number_of_samples > 100, 100, Number_of_samples)) %>% 
  ggplot(aes(Number_of_samples)) +
  geom_histogram(binwidth =1)+ggtitle("Number of samples with abundance larger than 0.05")+theme(plot.title = element_text(hjust = 0.5)) +
  theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title.x = element_text(size = 16),axis.title.y = element_text(size = 16))

plot(plot_sample_prevalence)
```

```{r restrict the species to those were over 0.05 in more than 10% samples, echo=FALSE}
#filter species that were over 0.05 in more than 10% samples
filter.species <- names(which(rowSums(metaphlan_species > 0.05) >= 27))
metaphlan_species <- metaphlan_species[filter.species,]

#re-normalization and arcsine transformation
metaphlan_species <- as.data.frame(t(apply(metaphlan_species, 2, function(x) x / sum(x)))) ##re-nomalize##
metaphlan_species_raw <- metaphlan_species ##raw data
metaphlan_species <- asin(sqrt(metaphlan_species))   ##Arcsine Transformation
```

```{r combine microbiome abundance with clinical data, warning=FALSE}
#match the specie abundance data with clinical data
metaphlan_species$sex <- clinical_data[match(rownames(metaphlan_species),clinical_data$Lab.ID),"sex"]
metaphlan_species$age <- clinical_data[match(rownames(metaphlan_species),clinical_data$Lab.ID),"age1"]
metaphlan_species$BMI <- clinical_data[match(rownames(metaphlan_species),clinical_data$Lab.ID),"calculated_bmi1"]
metaphlan_species$subclass <- clinical_data[match(rownames(metaphlan_species),clinical_data$Lab.ID),"subclass"]


#given that the case and control were defined by the BMI, therefore we can separate the two group by the median of BMI
metaphlan_species$Overweight <- NA
for (i in 1:nrow(metaphlan_species)) {
  metaphlan_species[i,"Overweight"] <- ifelse(metaphlan_species[i,"BMI"] > median(metaphlan_species$BMI),1,0)
}

###order the species by the number for the matching pairs
metaphlan_species <- metaphlan_species[order(metaphlan_species$subclass),]
```

```{r phyla data, warning=FALSE,fig.height=5, fig.width=7}
metaphlan_Phyla <- metaphlan_abundance[grep("p__",row.names(metaphlan_abundance)),]
metaphlan_Phyla <- metaphlan_Phyla[-grep("c__",row.names(metaphlan_Phyla)),]

row.names(metaphlan_Phyla) <- gsub("^.*p__","",row.names(metaphlan_Phyla))


metaphlan_Phyla <- metaphlan_Phyla[,grep("BC",colnames(metaphlan_Phyla))]

###exclude the unmatch pairs and samples failed QC
metaphlan_Phyla<- metaphlan_Phyla[,-which(colnames(metaphlan_Phyla) %in% c(failed_sample,unmatch.id))]

metaphlan_Phyla <- metaphlan_Phyla[-c(which(rowSums(metaphlan_Phyla)==0)),]



sample_prevalence <- as.data.frame(rowSums(metaphlan_Phyla>0.05))

colnames(sample_prevalence) <- "Number_of_samples"


sample_prevalence %>% 
  mutate(Number_of_samples = ifelse(Number_of_samples > 100, 100, Number_of_samples)) %>% 
  ggplot(aes(Number_of_samples)) +
  geom_histogram(binwidth =1)+ggtitle("Number of samples with abundance larger than 0.05")+theme(plot.title = element_text(hjust = 0.5)) +
  theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title.x = element_text(size = 16),axis.title.y = element_text(size = 16))

filter.Phyla <- names(which(rowSums(metaphlan_Phyla > 0.05) >= 27))



metaphlan_Phyla <- metaphlan_Phyla[filter.Phyla,]
metaphlan_Phyla <- as.data.frame(t(apply(metaphlan_Phyla, 2, function(x) x / sum(x)))) ##re-nomalize##

metaphlan_Phyla_raw <- metaphlan_Phyla ##raw data


metaphlan_Phyla <- asin(sqrt(metaphlan_Phyla))   ##Arcsine Transformation

metaphlan_Phyla$sex <- clinical_data[match(rownames(metaphlan_Phyla),clinical_data$Lab.ID),"sex"]
metaphlan_Phyla$age <- clinical_data[match(rownames(metaphlan_Phyla),clinical_data$Lab.ID),"age1"]
metaphlan_Phyla$BMI <- clinical_data[match(rownames(metaphlan_Phyla),clinical_data$Lab.ID),"calculated_bmi1"]



metaphlan_Phyla$subclass <- clinical_data[match(rownames(metaphlan_Phyla),clinical_data$Lab.ID),"subclass"]
metaphlan_Phyla$Overweight <- NA
for (i in 1:nrow(metaphlan_Phyla)) {
  metaphlan_Phyla[i,"Overweight"] <- ifelse(metaphlan_Phyla[i,"BMI"] > median(metaphlan_Phyla$BMI),1,0)
}
```

# Step 2: Data analysis for comparison of species and pathways between obesity and control

1.  **F/B Ratio**

```{r F/B ratio, warning=FALSE,fig.height=5, fig.width=8}
metaphlan_Phyla$F_B_ratio<-metaphlan_Phyla$Firmicutes/metaphlan_Phyla$Bacteroidetes
metaphlan_Phyla<- metaphlan_Phyla[order(metaphlan_Phyla$subclass),]

###try removing the extreme value to see if those extreme value affect the result
metaphlan_Phyla_testing<-metaphlan_Phyla%>%filter(!subclass %in% metaphlan_Phyla[which(metaphlan_Phyla$F_B_ratio>10),]$subclass)
metaphlan_Phyla_control <- subset(metaphlan_Phyla_testing,  Overweight == "0",drop = TRUE)
metaphlan_Phyla_case <- subset(metaphlan_Phyla_testing,  Overweight == "1",drop = TRUE)

beeplot(metaphlan_Phyla_control[,"F_B_ratio"],metaphlan_Phyla_case[,"F_B_ratio"],species = "F_B_ratio",group="Overweight",metaphlan_Phyla_testing, xlab="", ylab = "Ratio of Firmicutes to Bacteroidetes")

```

2.  **Top 15 species in obesity and control**

```{r Top 15 species in obesity and control,fig.height=5, fig.width=8, warning=FALSE}
library(pals)
# pal.bands(alphabet, alphabet2, cols25, glasbey, kelly, polychrome, 
#           stepped, tol, watlington,
#           show.names=FALSE)

alphabet2 <- colorRampPalette(colors=c("#AA0DFE", "#3283FE", "#85660D", "#782AB6", "#565656", 
                                                "#1C8356", "#16FF32", "#1CBE4F", 
                                                "#C4451C", "#DEA0FD", "#FE00FA", "#325A9B", "#FEAF16", 
                                                "#F8A19F", "#90AD1C", "#F6222E", "#1CFFCE", "#2ED9FF", 
                                                "#B10DA1", "#C075A6", "#FC1CBF", "#B00068", "#FBE426", 
                                                "#FA0087"))(24)
                                                


metaphlan_species_raw$Overweight <-  metaphlan_species[match(rownames(metaphlan_species_raw),rownames(metaphlan_species)),"Overweight"]

metaphlan_species_raw_obesity <- metaphlan_species_raw[which(metaphlan_species_raw$Overweight=="1"),]
metaphlan_species_raw_control <- metaphlan_species_raw[which(metaphlan_species_raw$Overweight=="0"),]  ###change for obesity and control

top_species_obesity <- data.frame(sort(apply(metaphlan_species_raw_obesity[,1:length(filter.species)],2,mean),decreasing = T)[1:15])
top_species_obesity$species <- row.names(top_species_obesity)

top_species_control <- data.frame(sort(apply(metaphlan_species_raw_control[,1:length(filter.species)],2,mean),decreasing = T)[1:15])
top_species_control$species <- row.names(top_species_control)

colnames(top_species_obesity)[1] <- "relative_abundance"
colnames(top_species_control)[1] <- "relative_abundance"

p1<-ggplot(data=top_species_control, aes(x=reorder(species, -relative_abundance), y=relative_abundance, color = reorder(species, -relative_abundance))) +
  geom_segment( aes(x=reorder(species, relative_abundance), xend=reorder(species, relative_abundance), y=0, yend=relative_abundance)) +
  geom_point(  size=4, alpha=0.7) +
  scale_color_manual(values = alphabet2)+
  theme_light() +
  coord_flip() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank(),
  )+labs(colour="Top 15 species")+xlab("")+ylab("Relative Abundance")+ggtitle("Control")+ guides(col="none")

p2<-ggplot(data=top_species_obesity, aes(x=reorder(species, -relative_abundance), y=relative_abundance, color = reorder(species, -relative_abundance))) +
  geom_segment( aes(x=reorder(species, relative_abundance), xend=reorder(species, relative_abundance), y=0, yend=relative_abundance)) +
  geom_point(  size=4, alpha=0.7) +
  scale_color_manual(values = alphabet2)+
  theme_light() +
  coord_flip() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank(),
  )+labs(colour="Top 15 species")+xlab("")+ylab("Relative Abundance")+ggtitle("Case")+ guides(col="none")

ggpubr::ggarrange(p1,p2,ncol=2,nrow=1)

```

3.  Alpha diversity

```{r Alpha diversity,fig.height=5, fig.width=7,  warning=FALSE}
Shannon_diversity(metaphlan_species_all)
Pielou_evenness(metaphlan_species_all)
simpson_diversity(metaphlan_species_all)


```

4.  **Beta diversity**

```{r Beta diversity (1)PCoA plot, fig.height=8, fig.width=8, warning=FALSE}

metaphlan_species$Overweight <- as.factor(metaphlan_species$Overweight)


bino_dist<-vegdist(metaphlan_species[,1:length(filter.species)],method = "bray")
species.pcoa <- pcoa(bino_dist)

df.plot<-data.frame(species.pcoa$vectors)
df.plot$Overweight <- metaphlan_species[match(rownames(df.plot),rownames(metaphlan_species)),"Overweight"]
df.plot$Overweight <- as.factor(df.plot$Overweight)
x_label<-round(species.pcoa$values$Rel_corr_eig[1]*100,2)
y_label<-round(species.pcoa$values$Rel_corr_eig[2]*100,2)
plot_pcoA<-ggplot(data=df.plot,aes(x=Axis.1,y=Axis.2,
                        color=Overweight,shape=Overweight))+
  geom_point(size=5)+
  theme_bw()+
  theme(panel.grid = element_blank())+
  geom_vline(xintercept = 0,lty="dashed")+
  geom_hline(yintercept = 0,lty="dashed")+
  labs(x=paste0("PCoA1 ",x_label,"%"),
       y=paste0("PCoA2 ",y_label,"%"))+
  stat_ellipse(data=df.plot,
               geom = "polygon",
               aes(fill=Overweight),
               alpha=0.3,level = 0.9)+
  scale_fill_manual(values = c("#e31a1c","#1f78b4"))

plot_pcoA
```

```{r Beta diversity (2) PERMANOVA, warning=FALSE}
set.seed(123)
pairwise.adonis(metaphlan_species[,1:length(filter.species)], metaphlan_species$Overweight,
                sim.method="bray", p.adjust.m= "fdr",perm = 9999)
```

```{r Beta diversity (3) PERMANOVA-FL, warning=FALSE}

testing.table<-metaphlan_species[,1:length(filter.species)]
testing.meta<-metaphlan_species[,c("sex","subclass","Overweight")]
res.perm.matched<-permanovaFL(formula=testing.table|as.factor(subclass)~Overweight,data =  testing.meta, 
                              seed=123,cluster.id = subclass,perm.within.type = "free",perm.between.type = "none",dist.method = "bray")

t(res.perm.matched)
```

5.  **Enrichment in species and pathways**

```{r  Enrichment in species and pathways, warning=FALSE,fig.height=5, fig.width=7}
#####################species enrichment

metaphlan_species_low <- subset(metaphlan_species,  Overweight == "0",drop = TRUE)
metaphlan_species_high <- subset(metaphlan_species,  Overweight == "1",drop = TRUE)


species_significant <- sapply(1:length(filter.species),function(i){wilcox.test(metaphlan_species_low[,i],metaphlan_species_high[,i],
                                                                               paired = T,alternative = "two.sided")$p.value})

colnames(metaphlan_species[which(species_significant<0.05)])
colnames(metaphlan_species[which(species_significant==min(species_significant))])
hist(species_significant,main = paste("Histogram of" , "P value"))

species_significant_adjusted <- p.adjust(species_significant, method = "fdr")
hist(species_significant_adjusted,main = paste("Histogram of" , "P.adjusted"))

colnames(metaphlan_species[which(species_significant_adjusted<0.15)])



for (species in colnames(metaphlan_species[which(species_significant_adjusted<0.15)])) {
beeplot(metaphlan_species_low[,species],metaphlan_species_high[,species],species,"Overweight",metaphlan_species,ylab=species)
}



#####################pathway enrichment

####data cleaning procedure for pathway
pathway_abundance <- fread("D://Annabel//OneDrive - The University Of Hong Kong//Study//postdoc//coprojects//Microbiome and obesity//sheng//path.txt")
pathway_abundance <- as.data.frame(pathway_abundance)
rownames(pathway_abundance) <- pathway_abundance$`# Pathway`; pathway_abundance <- pathway_abundance[,-1]

pathway_abundance <- pathway_abundance[-grep("UNINTEGRATED|UNMAPPED",rownames(pathway_abundance)),-grep("Coverage",colnames(pathway_abundance))]

colnames(pathway_abundance) <- gsub("_Abundance","",colnames(pathway_abundance))

pathway_abundance <- pathway_abundance[,which(colnames(pathway_abundance) %in% rownames(metaphlan_species))]

pathway_abundance_path <- pathway_abundance[-grep("\\|",rownames(pathway_abundance)),]

pathway_abundance_path <- pathway_abundance_path[-c(which(rowSums(pathway_abundance_path)==0)),]

filter.pathway <- names(which(rowSums(pathway_abundance_path > 0) >= 54))


pathway_abundance_path <- pathway_abundance_path[filter.pathway,]

pathway_abundance_path <- as.data.frame(t(apply(pathway_abundance_path, 2, function(x) x / sum(x)))) ##re-nomalize##


pathway_abundance_path_new <- pathway_abundance_path

pathway_abundance_path_new <-pathway_abundance_path_new[, order(colSums(pathway_abundance_path_new),decreasing = T)]

#colSums(pathway_abundance_path_new)

keep_going <- TRUE
i <- 1
while(keep_going){
  s <- seq(i+1, ncol(pathway_abundance_path_new))
  r <- cor(pathway_abundance_path_new[,s], pathway_abundance_path_new[,i])
  if(any(abs(r) > 0.70)){
    pathway_abundance_path_new <- pathway_abundance_path_new[, -s[which(abs(r) > 0.70)]]
  }
  i <- i+1
  if(ncol(pathway_abundance_path_new) <= i){
    keep_going <- FALSE
  }
}

pathway_abundance_path_new <- as.data.frame(t(apply(pathway_abundance_path_new, 1, function(x) x / sum(x)))) ##re-nomalize##

pathway_abundance_path_new$Overweight <- metaphlan_species[match(rownames(pathway_abundance_path_new),rownames(metaphlan_species)),"Overweight"]
pathway_abundance_path_new$subclass <- metaphlan_species[match(rownames(pathway_abundance_path_new),rownames(metaphlan_species)),"subclass"]
pathway_abundance_path_new <- pathway_abundance_path_new[order(pathway_abundance_path_new$subclass),]

###permanova


set.seed(123)
pairwise.adonis(pathway_abundance_path_new[,1:(ncol(pathway_abundance_path_new)-2)], pathway_abundance_path_new$Overweight,
                sim.method="binomial", p.adjust.m= "fdr",perm = 9999)

###permanova-FL


testing.table<-pathway_abundance_path_new[,1:(ncol(pathway_abundance_path_new)-2)]
testing.meta<-pathway_abundance_path_new[,c("subclass","Overweight")]
res.perm.matched<-permanovaFL(formula=testing.table|as.factor(subclass)~Overweight,data =  testing.meta, 
                              seed=123,cluster.id = subclass,perm.within.type = "free",perm.between.type = "none",dist.method = "binomial")

t(res.perm.matched)

##### pathway enrichment

pathway_low <- subset(pathway_abundance_path_new,  Overweight == "0",drop = TRUE)
pathway_high <- subset(pathway_abundance_path_new,  Overweight == "1",drop = TRUE)


pathway_abundance_path_sig <- sapply(1:(ncol(pathway_abundance_path_new)-2),function(i){wilcox.test(pathway_low[,i],pathway_high[,i],paired = T,alternative = "two.sided")$p.value})
#######plot all pathway

pdf(file="plots_path_sensitivity.pdf") 
par(mfrow= c(3,6),mar=c(1,4,1,1),mgp=c(3, 1.5, 0))
# for (pathway in colnames(pathway_abundance_path_new[which(pathway_abundance_path_sig<0.05)])) {
#     beeplot(pathway_low[,pathway],pathway_high[,pathway],pathway,"Overweight",pathway_abundance_path_new,ylab=pathway)
# }

dev.off() 



bino_dist<-vegdist(pathway_abundance_path_new[,1:(ncol(pathway_abundance_path_new)-2)],method = "binomial")
species.pcoa <- pcoa(bino_dist)

df.plot<-data.frame(species.pcoa$vectors)
df.plot$Overweight <- pathway_abundance_path_new[match(rownames(df.plot),rownames(pathway_abundance_path_new)),"Overweight"]
df.plot$Overweight <- as.factor(df.plot$Overweight)
x_label<-round(species.pcoa$values$Rel_corr_eig[1]*100,2)
y_label<-round(species.pcoa$values$Rel_corr_eig[2]*100,2)
ggplot(data=df.plot,aes(x=Axis.1,y=Axis.2,
                        color=Overweight,shape=Overweight))+
  geom_point(size=5)+
  theme_bw()+
  theme(panel.grid = element_blank())+
  geom_vline(xintercept = 0,lty="dashed")+
  geom_hline(yintercept = 0,lty="dashed")+
  labs(x=paste0("PCoA1 ",x_label,"%"),
       y=paste0("PCoA2 ",y_label,"%"))+
  stat_ellipse(data=df.plot,
               geom = "polygon",
               aes(fill=Overweight),
               alpha=0.3,level = 0.9)+
  scale_fill_manual(values = c("#e31a1c","#1f78b4"))
```

# Step 3: Conditional logistic regression- considering matching as well as binary outcome

```{r Conditional logistic regression,fig.height=12, fig.width=7, warning=FALSE}
########: 
########Conditional logistic regression- considering matching as well as binary outcome

################################################
#Analysis and transform the result into matrix##
################################################

#read other variable as possible confounders
other_variables_extracted<-read.csv("D:/Annabel/OneDrive - The University Of Hong Kong/Study/postdoc/coprojects/Microbiome and obesity/other_variables_extracted.csv")
other_variables_extracted<-other_variables_extracted %>%  mutate(c_sec=ifelse(deliv==3,1,0),
                                                                 ex_bf=ifelse(bf==2,1,0),
                                                                 preterm=ifelse(termbirth==1,0,1))
GDM<-read.csv("D:/Annabel/OneDrive - The University Of Hong Kong/Study/postdoc/coprojects/Microbiome and obesity/GDM.csv")
other_variables_extracted<-merge(GDM,other_variables_extracted,by="id")
Diet<-c("RC1","RC2","RC3","RC4")
Other_confounder<-c("msmk","deliv","bw_use","gestwkc2016","income","house","edup2016")
Other_confounder<-c("bw_use","gestwkc2016","income","house")


Other_confounder<-c("bf","parity","deliv","bw_use","gestwkc2016","income","house","edup2016")
library(dplyr)


diet_data <- read.csv("D:/Annabel/OneDrive - The University Of Hong Kong/Study/postdoc/coprojects/Microbiome and obesity/diet_information_extracted.csv")
rownames(diet_data) <- diet_data$Lab.ID
metaphlan_species$Lab.ID<-rownames(metaphlan_species)

Species_daiet_data <- merge(metaphlan_species,diet_data[,59:62],all.x = TRUE,by=0)
rownames(Species_daiet_data) <- Species_daiet_data$Lab.ID
Species_daiet_data <- Species_daiet_data[,-1]


Species_daiet_data<-Species_daiet_data %>%mutate(id=as.numeric(substr(row.names(Species_daiet_data),6,11)))
Species_daiet_data<-left_join(Species_daiet_data,other_variables_extracted[c("id",Other_confounder)],by="id")
Species_daiet_data$subclass<-as.integer(Species_daiet_data$subclass)
Species_daiet_data$Overweight<-as.numeric(Species_daiet_data$Overweight)

#since a majority of species doe not achieve the occurrence rate over 85%, they might be needed to transform into dichotomous traits(exist or not)
#when applying into general linear regression model like logistic regression.
Species_daiet_data <- merge(metaphlan_species,diet_data[,59:62],all.x = TRUE,by=0)
rownames(Species_daiet_data) <- Species_daiet_data$Row.names
Species_daiet_data <- Species_daiet_data[,-1]
Species_daiet_data<-Species_daiet_data %>%mutate(id=as.numeric(substr(row.names(Species_daiet_data),6,11)))
Species_daiet_data<-left_join(Species_daiet_data,other_variables_extracted[c("id",Other_confounder)],by="id")
bac_log.z<-scale(log1p(Species_daiet_data[filter.species]))#for standardization
bac_exist<-Species_daiet_data[filter.species]>0

#Check with the probability of each species exist
check<-sapply(1:ncol(bac_exist),function(x){prop.table(table(bac_exist[,x]))})
hist(check[2,])
table(check[2,]>0.85)#30 out of 192
which(check[2,]>0.85)
table(check[2,]>0.90)#15 out of 192
table(check[2,]>0.95)#9 out of 192


DATA_analysis<-cbind(bac_log.z[,which(check[2,]>0.85)],bac_exist[,which(!check[2,]>0.85)],Species_daiet_data[,193:(length(Species_daiet_data))])

colnames(bac_log.z[,which(check[2,]>0.85)])
colnames(bac_exist[,which(!check[2,]>0.85)])


###DATA analysis
condi_logi1<-condi_logi2<- condi_logi3<-list(NA)
DATA_analysis$subclass<-as.integer(DATA_analysis$subclass)
DATA_analysis$Overweight<-as.numeric(DATA_analysis$Overweight)

for (i in 1:length(filter.species)) {
  model1 <-as.formula(paste("Overweight ~",filter.species[i],"+strata(subclass)",sep=""))
  model2 <-as.formula(paste("Overweight ~",filter.species[i],"+strata(subclass)+",paste(Other_confounder,collapse = "+"),sep=""))
  model3 <-as.formula(paste("Overweight ~",filter.species[i],"+strata(subclass)+",paste(Diet[1:4],collapse = "+"),"+",paste(Other_confounder,collapse = "+"),sep=""))
  condi_logi1[[i]]<-rbind(summary(clogit(model1,data=DATA_analysis))$coef[1,],
                          summary(clogit(model1,data=DATA_analysis[which(DATA_analysis$sex=="F"),]))$coef[1,],
                          summary(clogit(model1,data=DATA_analysis[which(DATA_analysis$sex=="M"),]))$coef[1,])%>%
                    as_tibble()%>%mutate(bac=filter.species[i],Gender=c("All","Girls","Boys"),model="m1")
  condi_logi2[[i]]<-rbind(summary(clogit(model2,data=DATA_analysis))$coef[1,],
                          summary(clogit(model2,data=DATA_analysis[which(DATA_analysis$sex=="F"),]))$coef[1,],
                          summary(clogit(model2,data=DATA_analysis[which(DATA_analysis$sex=="M"),]))$coef[1,])%>%
                    as_tibble()%>%mutate(bac=filter.species[i],Gender=c("All","Girls","Boys"),model="m2")
   condi_logi3[[i]]<-rbind(summary(clogit(model3,data=DATA_analysis))$coef[1,],
                           summary(clogit(model3,data=DATA_analysis[which(DATA_analysis$sex=="F"),]))$coef[1,],
                           summary(clogit(model3,data=DATA_analysis[which(DATA_analysis$sex=="M"),]))$coef[1,])%>%
                    as_tibble()%>%mutate(bac=filter.species[i],Gender=c("All","Girls","Boys"),model="m3")
  
}

bind_model1<-do.call(rbind,condi_logi1)
bind_model2<-do.call(rbind,condi_logi2)
bind_model3<-do.call(rbind,condi_logi3)

cor.mat<-rbind(bind_model1,bind_model2,bind_model3)
cor.mat2<-cor.mat %>% filter(Gender=="All")

cor.mat<-cor.mat %>%
       group_by(Gender,model) %>% 
       mutate(pval.adj = p.adjust (`Pr(>|z|)`, method='fdr'))%>%as.data.frame()

cor.mat<-cor.mat2 %>%
  group_by(model) %>% 
  mutate(pval.adj = p.adjust (`Pr(>|z|)`, method='fdr'))%>%as.data.frame()

cor.mat$pval <- as.numeric(cor.mat$`Pr(>|z|)`)
filt.feat <- cor.mat[cor.mat$pval<=0.01,"bac"]

cor.mat <- cor.mat[cor.mat$bac%in%unique(filt.feat),]
cor.mat$sym <- ifelse(cor.mat$pval>=0.05, "", ifelse(cor.mat$pval<=0.01,ifelse(cor.mat$pval.adj<0.15,"***", "**"), "*"))
cor.mat$direction <-ifelse(cor.mat$coef>0,+1,-1)
cor.mat$direction <- as.factor(cor.mat$direction)



#check the units for each species
unique(cor.mat$bac)

unique(cor.mat$bac)[which(unique(cor.mat$bac)%in% colnames(bac_log.z[,which(check[2,]>0.85)]))]
# "Bacteroides_thetaiotaomicron"  "Bacteroides_uniformis"         "Flavonifractor_plautii"        "Phascolarctobacterium_faecium"


####################################################################################################################

#Add Estimates
`cor.mat`<-cor.mat%>% mutate(CI_95upper=exp(coef+1.96*`se(coef)`),
                           CI_95lower=exp(coef-1.96*`se(coef)`),
                           OR=round(`exp(coef)`,2))%>%
  mutate(Estimate=paste(format(OR,nsmall=2,trim=T),"(",format(round(CI_95lower,2),nsmall=2,trim=T),",",format(round(CI_95upper,2),nsmall=2,trim=T),")",sep=""))

write.csv(cor.mat,"CLRM_microbiome_Overweight_m1-m3_p0.01.csv",row.names = F)

p2<-ggplot(cor.mat, aes(x=model, y = bac, fill=OR)) + geom_tile(color="white", size=0.1)+
  coord_equal() + xlab("") + ylab("") + geom_text(aes(label=paste(format(OR,nsmall=2,trim=T),sym,sep = "")), size = 3, col = "black")+ #facet_wrap(~Gender) +
  theme(axis.text = element_text(size = 10, face="bold"), strip.text = element_text(size=8, face="bold"),
        legend.text = element_text(size = 12, face="bold"), legend.title = element_text(size = 12, face="bold"),
        axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1))+ggtitle("")+
  # scale_fill_manual(name = "Species enrichment", 
  #                   labels = c("-1" = "Control", 
  #                              "1" = "Overweight"),
  #                   values = c("-1" = "#d4b4f6", 
  #                              "1" = "#9597f0"))+ 
  scale_fill_gradient2(low = "darkblue",high="darkred",guide = "colorbar",trans="log",limits=round(c(0.05,15),2),breaks=c(0.10,1.00,8.00))+
theme_minimal()+
  labs(caption = 
"m1 - crude model considering matching;                                     
m2 - m1 + adjusted for income, housing, birth weight and gestational age at birth;  
m3 - m2 + adjusted for top 4 principal components of diet information;               
* - p < 0.05;                                                                        
** - p < 0.01;                                                                       
*** - FDR<0.15;                                                   

a. Continuous data of the relative abundance after arcsine square root transformation,
   log-transformation and normalization (Units: SD) was used for Bacteroides uniformis
    Flavonifractor plautii, and Phascolarctobacterium faecium;
b. Binary data of existance comparing with non-exists was used for the other species
   with existance rate<85%.")+
  theme(plot.caption = element_text(hjust=0))



p2+ theme(plot.margin = unit(c(0,0,0,-5), "cm"))# the figure would be similar for the analysis using ranking or using continuous / existence according to existence <0.85




```

# Step 4: MR Analysis

The data for this analyis were obtain from the public available GWAS data from 4D-SZ, AGEN and Biobank Japan.

```{r, warning=FALSE}
library(ieugwasr)
options(ieugwasr_api = 'gwas-api.mrcieu.ac.uk/')
library(TwoSampleMR)
setwd("D:\\Annabel\\OneDrive - The University Of Hong Kong\\backup D\\Annabel\\MR\\consortium\\4D_SZ\\")
species<-list(NA)
# genus<-list(NA)
species[[1]]<-fread("Bacteroides_uniformis.csv")
species[[2]]<-fread("Clostridium_spiroforme.csv")


species<-lapply(species, function(x){x%>% rename(SNP=rsid,
                                             effect_allele=A1,
                                             other_allele=A2,
                                             beta=b,
                                             se=se,
                                             eaf=freq,
                                             pval=p)})

species[[1]]$exposure<-"s_Bacteroides_uniformis"
species[[2]]$exposure<-"s_Clostridium_spiroforme"


species<-lapply(species, as.data.frame)
exposure_gwas<-species
exp_dat<-list(NA)

exp_dat<-lapply(exposure_gwas, format_ins_select)

sapply(exp_dat, nrow)

####1.AGEN(T2D)##########################################
out_dat<-list(NA)
out_dat<-lapply(1:length(exp_dat), function(x){extract_outcome_data(exp_dat[[x]]$SNP,outcomes = "ebi-a-GCST010118")})
MR_custom(exp_dat,out_dat,outcome_gwas_name ="GWAS2020_T2D")



###2. BBJ-CAD##########################################
out_dat<-list(NA)
out_dat<-lapply(1:length(exp_dat), function(x){extract_outcome_data(exp_dat[[x]]$SNP,outcomes = "bbj-a-159")})
 MR_custom(exp_dat,out_dat,outcome_gwas_name ="bbj_CAD")


```
